<launch>
    <!-- Define arguments with default values -->
    <arg name="image_topic" default="/fisheye/image_raw" />
    <arg name="pose_topic" default="/orb_slam2_mono/pose" />
    <arg name="map_topic" default="/orb_slam2_mono/map" />
    <arg name="rendered_image_topic" default="/orb_slam2_mono/rendered_image" />
    <arg name="orb_slam2_parameter_file" default="slam_base.yaml" /> <!-- fisheye_adaptor will provide camera intrinsic as a topic -->
    <arg name="fisheye_calibration_file" default="fisheye_starling2_front.yaml" />
    <arg name="resize_factor" default="2.0" />

    <arg name="load_map" default="false" />
    <arg name="map_file_name" default="map.bin" />

    <arg name="publish_map" default="true" />
    <arg name="publish_rendered_image" default="true" />
    <arg name="publish_pose" default="true" />
    <arg name="publish_tf" default="false" />

    <node name="fisheye_adaptor" pkg="orb_slam2_ros" type="orb_slam2_ros_mono_fisheye_adaptor" output="screen">
        <!-- topic remapping -->
        <remap from="/fisheye/image_raw" to="$(arg image_topic)" />

        <!-- static parameters -->
        <param name="calibration_file" type="string" value="$(find orb_slam2_ros)/config/$(arg fisheye_calibration_file)" />
        <param name="resize_factor" type="double" value="$(arg resize_factor)" />
    </node>
    <node name="orb_slam2_mono" pkg="orb_slam2_ros" type="orb_slam2_ros_mono" output="screen">
        <!-- topic remapping -->
        <remap from="/camera/image_raw" to="/fisheye_adaptor/image_undistorted" />
        <remap from="/orb_slam2_mono/pose" to="$(arg pose_topic)" />
        <remap from="/orb_slam2_mono/map" to="$(arg map_topic)" />
        <remap from="/orb_slam2_mono/rendered_image" to="$(arg rendered_image_topic)" />

        <!-- static parameters -->
        <param name="publish_map" type="bool" value="$(arg publish_map)" />
        <param name="publish_rendered_image" type="bool" value="$(arg publish_rendered_image)" />
        <param name="publish_pose" type="bool" value="$(arg publish_pose)" />
        <param name="publish_tf" type="bool" value="$(arg publish_tf)" />

        <param name="load_map" type="bool" value="$(arg load_map)" />
        <param name="map_file_name" type="string" value="$(find orb_slam2_ros)/resource/$(arg map_file_name)" />

        <param name="vocabulary_file_name" type="string" value="$(find orb_slam2_ros)/orb_slam2_core/Vocabulary/ORBvoc.txt" />
        <param name="parameter_file" type="string" value="$(find orb_slam2_ros)/config/$(arg orb_slam2_parameter_file)" />
        <param name="camera_info_topic" type="string" value="/fisheye_adaptor/rectified_camera_info" />
    </node>
</launch>